In [3]:
import pandas as pd


In [4]:
listings = pd.read_csv('../data/raw/bangkok/listings.csv.gz', compression='gzip', low_memory=False)

In [5]:
listings.shape

(28806, 79)

In [7]:
listings.head ()

,id,listing_url,scrape_id,last_scraped,source,name,description,neighborhood_overview,picture_url,host_id,...,review_scores_communication,review_scores_location,review_scores_value,license,instant_bookable,calculated_host_listings_count,calculated_host_listings_count_entire_homes,calculated_host_listings_count_private_rooms,calculated_host_listings_count_shared_rooms,reviews_per_month
0,27934,https://www.airbnb.com/rooms/27934,20250926165947,2025-09-27,city scrape,Nice room with superb city view,Our cool and comfortable one bedroom apartment...,It is very center of Bangkok and easy access t...,https://a0.muscache.com/pictures/566374/23157b...,120437,...,4.91,4.66,4.75,NaN,f,1,1,0,0,0.40
1,27979,https://www.airbnb.com/rooms/27979,20250926165947,2025-09-27,previous scrape,"Easy going landlord,easy place",NaN,NaN,https://a0.muscache.com/pictures/106247594/1d6...,120541,...,NaN,NaN,NaN,NaN,f,2,1,1,0,NaN
2,28745,https://www.airbnb.com/rooms/28745,20250926165947,2025-09-27,previous scrape,modern-style apartment in Bangkok,A modern-style apartment situated in a cool ne...,Ramkumheang neigbourhood - hip place with food...,https://a0.muscache.com/pictures/160369/92a7ec...,123784,...,NaN,NaN,NaN,NaN,f,1,0,1,0,NaN
3,47516,https://www.airbnb.com/rooms/47516,20250926165947,2025-09-27,city scrape,Beautiful waterfront house,NaN,NaN,https://a0.muscache.com/pictures/271998/b090ad...,214456,...,NaN,NaN,NaN,NaN,f,1,1,0,0,NaN
4,48736,https://www.airbnb.com/rooms/48736,20250926165947,2025-09-27,city scrape,Condo with Chaopraya River View,at a size of 25sqm for your vacation in BKK.<b...,NaN,https://a0.muscache.com/pictures/12465296/e9c0...,222005,...,5.00,5.00,5.00,NaN,f,1,0,1,0,0.01


In [8]:
listings.columns

Index(['id', 'listing_url', 'scrape_id', 'last_scraped', 'source', 'name',
       'description', 'neighborhood_overview', 'picture_url', 'host_id',
       'host_url', 'host_name', 'host_since', 'host_location', 'host_about',
       'host_response_time', 'host_response_rate', 'host_acceptance_rate',
       'host_is_superhost', 'host_thumbnail_url', 'host_picture_url',
       'host_neighbourhood', 'host_listings_count',
       'host_total_listings_count', 'host_verifications',
       'host_has_profile_pic', 'host_identity_verified', 'neighbourhood',
       'neighbourhood_cleansed', 'neighbourhood_group_cleansed', 'latitude',
       'longitude', 'property_type', 'room_type', 'accommodates', 'bathrooms',
       'bathrooms_text', 'bedrooms', 'beds', 'amenities', 'price',
       'minimum_nights', 'maximum_nights', 'minimum_minimum_nights',
       'maximum_minimum_nights', 'minimum_maximum_nights',
       'maximum_maximum_nights', 'minimum_nights_avg_ntm',
       'maximum_nights_avg_ntm', 'ca

In [9]:
listings.info()

<class 'pandas.DataFrame'>
RangeIndex: 28806 entries, 0 to 28805
Data columns (total 79 columns):
 #   Column                                        Non-Null Count  Dtype  
---  ------                                        --------------  -----  
 0   id                                            28806 non-null  int64  
 1   listing_url                                   28806 non-null  str    
 2   scrape_id                                     28806 non-null  int64  
 3   last_scraped                                  28806 non-null  str    
 4   source                                        28806 non-null  str    
 5   name                                          28806 non-null  str    
 6   description                                   28030 non-null  str    
 7   neighborhood_overview                         9484 non-null   str    
 8   picture_url                                   28806 non-null  str    
 9   host_id                                       28806 non-null  int64  
 1

In [10]:
listings['price'].head(10)

0    $1,595.00
1          NaN
2          NaN
3    $4,188.00
4    $1,450.00
5    $1,368.00
6          NaN
7    $5,600.00
8    $1,147.00
9    $1,416.00
Name: price, dtype: str

In [11]:
listings['price'].str.replace('$', '', regex=False).str.replace(',', '', regex=False).head(10)

0    1595.00
1        NaN
2        NaN
3    4188.00
4    1450.00
5    1368.00
6        NaN
7    5600.00
8    1147.00
9    1416.00
Name: price, dtype: str

In [12]:
listings['price_clean'] = listings['price'].str.replace('$', '', regex=False).str.replace(',', '', regex=False).astype(float)

In [13]:
listings[['price', 'price_clean']].head(10)

,price,price_clean
0,"$1,595.00",1595.0
1,NaN,NaN
2,NaN,NaN
3,"$4,188.00",4188.0
4,"$1,450.00",1450.0
5,"$1,368.00",1368.0
6,NaN,NaN
7,"$5,600.00",5600.0
8,"$1,147.00",1147.0
9,"$1,416.00",1416.0


In [14]:
listings['price_clean'].describe()

count      23273.000000
mean        2528.749151
std        16473.896035
min            4.000000
25%          923.000000
50%         1379.000000
75%         2207.000000
max      1000000.000000
Name: price_clean, dtype: float64

In [15]:
listings[listings['price_clean'] > 50000][['name', 'price_clean', 'room_type']]

,name,price_clean,room_type
979,"Modern,wifi,5m MRT&2 Shopping Mall",1000000.0,Entire home/apt
1442,"2 Bedroom,wifi,5mMRT&Shopping Mall",1000000.0,Entire home/apt
1575,Luxury Duplex 68 Sqm. PC+WiFi 50 Mbps.80 M.to BTS,54442.0,Entire home/apt
1946,"Resort Style Luxury apartment,5min to MRT,free...",1000000.0,Entire home/apt
2130,"BTS Wongwianyai, Rooftop Pool, Wifi",152000.0,Entire home/apt
2578,New Luxury Duplex 60 Sqm.Jacuzzi WiFi 80m.to BTS,53633.0,Entire home/apt
2635,Luxury 7 Bed Rooms Swimming Pool near Sathorn ...,88000.0,Entire home/apt
2977,Luxury 2 Bedroom apartment close to everthing,70886.0,Private room
3276,Ideo verse ratchaprarop condo,928572.0,Private room
3723,"Modern&Luxury apartment,5min to MRT,freeWifi",1000000.0,Entire home/apt


In [16]:
listings[listings['price_clean'] == 1000000]

,id,listing_url,scrape_id,last_scraped,source,name,description,neighborhood_overview,picture_url,host_id,...,review_scores_location,review_scores_value,license,instant_bookable,calculated_host_listings_count,calculated_host_listings_count_entire_homes,calculated_host_listings_count_private_rooms,calculated_host_listings_count_shared_rooms,reviews_per_month,price_clean
979,8880974,https://www.airbnb.com/rooms/8880974,20250926165947,2025-09-27,city scrape,"Modern,wifi,5m MRT&2 Shopping Mall",Aspire Condo 37sqm 1 bedroom condominium. just...,Phra Ram 9 is a lively area conveniently locat...,https://a0.muscache.com/pictures/c71639a0-d9d4...,30803267,...,4.87,4.80,NaN,f,4,4,0,0,0.37,1000000.0
1442,11579635,https://www.airbnb.com/rooms/11579635,20250926165947,2025-09-27,city scrape,"2 Bedroom,wifi,5mMRT&Shopping Mall",You will feel comfortable and at home while st...,Phra ram 9 is a lively area conveniently locat...,https://a0.muscache.com/pictures/770ac184-8d94...,30803267,...,4.60,4.53,NaN,f,4,4,0,0,0.46,1000000.0
1946,15380964,https://www.airbnb.com/rooms/15380964,20250926165947,2025-09-27,city scrape,"Resort Style Luxury apartment,5min to MRT,free...",Our large and comfortable 2 bedroom apartment(...,Phra Ram 9 is a lively area conveniently locat...,https://a0.muscache.com/pictures/4f9a183d-ee5e...,30803267,...,4.73,4.23,NaN,f,4,4,0,0,0.38,1000000.0
3723,23740993,https://www.airbnb.com/rooms/23740993,20250926165947,2025-09-27,city scrape,"Modern&Luxury apartment,5min to MRT,freeWifi",Our large and comfortable 2 bedroom apartment(...,NaN,https://a0.muscache.com/pictures/06e326da-c032...,30803267,...,4.75,4.67,NaN,f,4,4,0,0,0.14,1000000.0


In [17]:
listings['price_flag_suspicious'] = listings['price_clean'] == 1000000
listings['price_flag_suspicious'].sum()

np.int64(4)

In [18]:
listings_for_analysis = listings[listings['price_flag_suspicious'] == False].copy()
listings_for_analysis['price_clean'].describe()

count     23269.000000
mean       2357.281318
std       10018.229296
min           4.000000
25%         923.000000
50%        1379.000000
75%        2207.000000
max      928572.000000
Name: price_clean, dtype: float64

In [19]:
listings['host_response_rate'].head(10)

0     NaN
1     NaN
2     NaN
3     NaN
4     NaN
5     67%
6     NaN
7      0%
8     NaN
9    100%
Name: host_response_rate, dtype: str

In [20]:
listings['host_response_rate'].isnull().sum()

np.int64(5647)

In [21]:
listings['host_response_rate'].str.replace('%', '', regex=False).astype(float).head(10)

0      NaN
1      NaN
2      NaN
3      NaN
4      NaN
5     67.0
6      NaN
7      0.0
8      NaN
9    100.0
Name: host_response_rate, dtype: float64

In [22]:
listings['host_response_rate_clean'] = listings['host_response_rate'].str.replace('%', '', regex=False).astype(float)
listings['host_acceptance_rate_clean'] = listings['host_acceptance_rate'].str.replace('%', '', regex=False).astype(float)

In [23]:
listings[['host_response_rate', 'host_response_rate_clean', 'host_acceptance_rate', 'host_acceptance_rate_clean']].head(10)

,host_response_rate,host_response_rate_clean,host_acceptance_rate,host_acceptance_rate_clean
0,NaN,NaN,67%,67.0
1,NaN,NaN,NaN,NaN
2,NaN,NaN,NaN,NaN
3,NaN,NaN,NaN,NaN
4,NaN,NaN,100%,100.0
5,67%,67.0,68%,68.0
6,NaN,NaN,NaN,NaN
7,0%,0.0,0%,0.0
8,NaN,NaN,NaN,NaN
9,100%,100.0,91%,91.0


In [24]:
neighbourhoods = pd.read_csv('../data/raw/bangkok/neighbourhoods.csv')
neighbourhoods.shape

(50, 2)

In [25]:
neighbourhoods.head(10)

,neighbourhood_group,neighbourhood
0,NaN,Bang Bon
1,NaN,Bang Kapi
2,NaN,Bang Khae
3,NaN,Bang Khen
4,NaN,Bang Kho laen
5,NaN,Bang Khun thain
6,NaN,Bangkok Noi
7,NaN,Bangkok Yai
8,NaN,Bang Na
9,NaN,Bang Phlat


In [26]:
neighbourhoods['neighbourhood_group'].isnull().sum()

np.int64(50)

In [28]:
reviews = pd.read_csv('../data/raw/bangkok/reviews.csv.gz', compression='gzip')
reviews.shape

(583333, 6)

In [29]:
reviews.head()

,listing_id,id,date,reviewer_id,reviewer_name,comments
0,27934,1094339,2012-04-07,1368195,Michael,We stayed in the apartment for a week and we e...
1,27934,1241042,2012-05-07,2007324,Scott,My girlfriend and I recently stayed in Nuttee'...
2,27934,1523384,2012-06-20,2263352,Marc,I stayed for one month at the condo and was re...
3,27934,1655571,2012-07-08,558987,Leyla,Nuttee was a great host! I really enjoyed her ...
4,27934,1972192,2012-08-13,2359865,Rachel,Nuttee was an amazing host. She and her daught...


In [30]:
reviews.info()

<class 'pandas.DataFrame'>
RangeIndex: 583333 entries, 0 to 583332
Data columns (total 6 columns):
 #   Column         Non-Null Count   Dtype
---  ------         --------------   -----
 0   listing_id     583333 non-null  int64
 1   id             583333 non-null  int64
 2   date           583333 non-null  str  
 3   reviewer_id    583333 non-null  int64
 4   reviewer_name  583327 non-null  str  
 5   comments       583263 non-null  str  
dtypes: int64(3), str(3)
memory usage: 26.7 MB


In [31]:
reviews['listing_id'].isin(listings['id']).value_counts()

listing_id
True    583333
Name: count, dtype: int64

In [32]:
listings['neighbourhood_cleansed'].isin(neighbourhoods['neighbourhood']).value_counts()

neighbourhood_cleansed
True    28806
Name: count, dtype: int64

In [33]:
calendar = pd.read_csv('../data/raw/bangkok/calendar.csv.gz', compression='gzip')
calendar.shape

(10514202, 7)

In [34]:
calendar.head()

,listing_id,date,available,price,adjusted_price,minimum_nights,maximum_nights
0,5675191,2025-09-27,f,NaN,NaN,4,1125
1,1920431,2025-09-26,f,NaN,NaN,1,31
2,1920431,2025-09-27,t,NaN,NaN,1,31
3,1920431,2025-09-28,f,NaN,NaN,1,31
4,1920431,2025-09-29,t,NaN,NaN,1,31


In [35]:
calendar.info()

<class 'pandas.DataFrame'>
RangeIndex: 10514202 entries, 0 to 10514201
Data columns (total 7 columns):
 #   Column          Dtype  
---  ------          -----  
 0   listing_id      int64  
 1   date            str    
 2   available       str    
 3   price           float64
 4   adjusted_price  float64
 5   minimum_nights  int64  
 6   maximum_nights  int64  
dtypes: float64(2), int64(3), str(2)
memory usage: 561.5 MB


In [36]:
calendar.isnull().sum()

listing_id               0
date                     0
available                0
price             10514202
adjusted_price    10514202
minimum_nights           0
maximum_nights           0
dtype: int64

In [37]:
calendar[calendar['available'] == 't']['price'].isnull().sum()

np.int64(7205796)

In [38]:
available_rows = (calendar['available'] == 't').sum()
available_rows

np.int64(7205796)

In [39]:
calendar[calendar['available'] == 't']['price'].notnull().sum()

np.int64(0)

**Business Domain Context**

**Listing:** A bookable accommodation unit on Airbnb - the core unit of supply in the marketplace. Each listing belongs to exactly one host and sits within one neighbourhood. In this dataset, a listing is uniquely identified by `id`.

**Host:** An individual or commercial operator who owns/manages one or more listings. We found 8,874 unique hosts behind 28,806 listings - meaning many hosts run multiple properties (e.g. some hosts have 16+ listings, indicating professional/commercial-scale operations rather than casual single-room hosts).

- **Review**: Demand-side feedback left by a guest after a completed stay. A review generally implies a completed booking, making review volume a useful (if imperfect) proxy for booking activity. We confirmed all 583,333 reviews correctly link back to a real listing.

- **Neighbourhood**: A geographic district within Bangkok (50 total). Used to group listings spatially for pricing and density analysis. Bangkok has a flat structure - no borough/group-level hierarchy exists in this data.

- **Calendar day**: A single listing's availability state for one specific future date. In this dataset, the calendar's `price` field was found to be completely empty (100% missing), so it cannot be used for date-level pricing - availability/revenue analysis should instead rely on the pre-aggregated fields already present in `listings.csv.gz` (e.g. `availability_365`, `estimated_revenue_l365d`).

**Assumptions & Decisions Log - Dataset Familiarization**

| Decision | What I Found | What I Did | Why |
|---|---|---|---|
| Currency interpretation | `price` field shows `$` symbol (e.g. `$1,595.00`) | Treated as Thai Baht (THB), not USD | Inside Airbnb's documented convention for non-US cities is to report local currency, despite the `$` formatting artifact |
| Empty columns | `license`, `calendar_updated`, `neighbourhood_group_cleansed` are 100% null (0/28,806) | Will exclude from analysis/modeling | Zero non-null values means zero information content |
| Suspicious price outliers | 4 listings priced at exactly 1,000,000 (a suspiciously round number) | Flagged via new `price_flag_suspicious` column, excluded from descriptive statistics, but not deleted from raw data | Exact round-number values strongly suggest placeholder/error data, not real prices; flagging preserves transparency over silent deletion |
| Legitimate luxury listings | Listings priced 50,000–928,572 THB exist (villas, penthouses, multi-bedroom properties) | Kept in the dataset | These represent a real luxury market segment; removing them would bias the analysis by hiding a legitimate part of the market |
| `neighbourhood` vs `neighbourhood_cleansed` | `neighbourhood` is 67% null/free-text; `neighbourhood_cleansed` is 100% complete and standardized | Will always use `neighbourhood_cleansed` for grouping/analysis | `neighbourhood_cleansed` is geocoded and validated; `neighbourhood` is unreliable host-entered text |
| Calendar price data | 100% of `calendar.csv.gz` price/adjusted_price values are null, even on available days (verified: 0 of 7,205,796 available-day rows have a price) | Will not use calendar for date-level pricing; will rely on `listings.csv.gz` pre-aggregated fields instead (`availability_365`, `estimated_revenue_l365d`) | Calendar pricing data was not captured in this Bangkok scrape - a genuine dataset limitation, not something fixable through cleaning |
| Mean vs median for "typical price" | Mean (2,528 THB) is much higher than median (1,379 THB) before cleaning | Will favor median for describing "typical" listings; report both with explanation when relevant | Large gap between mean and median indicates a right-skewed distribution driven by a small number of extreme high-price listings |